### ISSUE TITLE
qapply re-write: full evaluation, no expand(), 

### ISSUE THREAD

**Issue by Costor** [sympy/sympy#24348]
### Situation ###
As discussed in https://github.com/sympy/sympy/discussions/24159 (also see excerpt below), I have done a re-write of qapply. 

### Issues (fixed) ###

In order of relevance the various issues dealt with were:
1. qapply no longer may stop short in computing expressions, especially involving dyads and or powers of composite expressions, returning unnecessarily complex results only partially evaluated
2. qapply no longer calls expand() during computation, which fixes issue #6268. Users may invoke expand(tensorproduct=True) on the result at their discretion instead
3. qapply now iterates through long factor lists and powers insted of doing recursion which could make it hit Python's recursion limit when computing something like `(some-gate)**2000 * Qubit(0)`
4. qapply now has capabilities to work with symbolic and fractional powers and tries some simplifications to reduce roll of of powers

Examples for 1.:
```
from sympy.physics.quantum.qubit import *
from sympy.physics.quantum.gate import *
from sympy.physics.quantum.qapply import *
print(qapply((Qubit(1)*QubitBra(1)) * (QubitBra(1)*Qubit(1)))) # -> <1|1> |1><1|  instead of |1><1|
print(qapply(QubitBra(1)*QubitBra(1)*Qubit(1)*X(0), dagger=True)) # -> <1|*X(0) instead of <0|
print(qapply(QubitBra(1)*QubitBra(1)*Qubit(1)*Qubit(1), dagger=True)) # -> <1|**2 * |1>**2 instead of 1
dyad11 = Qubit(1)*QubitBra(1)
print(qapply(dyad11*dyad11*dyad11)) # -> dyad11**3 instead of dyad11
```
Especially 1. limited the usability of qapply to compute sequences of controlled gates and other terms that appear e.g. in quantum computing, as the effects tend to accumulate quickly in larger expressions. When I tried to fix this in the existing code basis I found it needed a more explicit design how to identify factors and applying factors to each other. The same was true for 2.

### Result ###
The result is essentially a generic expression evaluator for objects of a non-commutative operator algebra which knows  about Mul and Add. How factors act on each other is provided by methods of the objects, not by qapply. The only explicit knowledge about quantum objects that had to be hardcoded is how to break them up into their factors (if any) or how to call their constructors. The same could be done for any other family of objects, e.g. operators, matrices and vectors from Linear Algebra.

### Effects ###

As qapply no longer calls expand() regularly, the result of qapply will be unexpanded. They may need expansion for two reasons: i) either the result is compared against ii) the result would simplify from expansion as summands cancel out  (e.g.  a*b + a*(c-b) -> a*c.  Both effects may lead to unexpected results and/or strong increase in run time, so it is advisable to review invocations of qapply and to decide whether the addition of an .expand() or .expand(tensorproduct=True) is required. 

### Steps ###

I'll provide a PR for qapply plus additional test cases.

### Excerpt from https://github.com/sympy/sympy/discussions/24159

<div type='discussions-op-text'>

<sup>Originally posted by **Costor** October 22, 2022</sup>
My background is math, computer science and quantum computing.  I've just dug into qapply of the quantum package stumbling over issues #24142, #24153 and #24158 (and more unpublished). I also provided fixes for these, and also did a re-write of qapply for my purposes along the way.

Three questions: 

1. I've also seen (e.g. in #6268) that there might be reasonable concern about changing the behaviour of qapply. So it is worth while to continue a re-write? What would be conditions for a re-write to be accepted?
5. I feel there is no counterpart to qapply in the linear algebra packages of SymPy that would work with e.g. MatrixSymbol, generic vectors etc. the symbolic way qapply can with the quantum objects. Is that correct? If yes - what about thinking to carry over qapply to these objects too - actually both packages do linear algebra?
6. Pitfalls you can think of?

### How did I get there ###

Some days ago I wanted to do some quick computation of some controlled gates, i.e. compute a product of four tensor product operators (including OuterProduct |1><1| in some TensorProducts) with SymPy. 
I couldn't get a result from qapply and have already filed issues #24142, #24153 and #24158.  

I also found that qapply needed improvement in handling OuterProducts, powers of operators and some more. When I tried to fix this in the existing code I didn't see how to do it there.

So I re-wrote qapply as an generic expression evaluator, with a generic core that only knows about Add and Mul and is driven by properties and methods of the expression elements. By construction it also solves #6268 as it just follows the expression tree and distributes all forms of factors over all forms of sums explicitly without ever calling .expand(). I also limited recursion to the depth of the expression tree to avoid that larger powers or numbers of factors hit Python's recursion limit.

That sufficed for my initial computation task at hand, and it runs all test cases in test_qapply (and many more, with the exception of an additional expand() of the result), but of course it is fresh unproven code with some features missing. Nevertheless I consider it a reasonable basis as I feel the structure is clearer.

</div>

**Comment by oscarbenjamin** [comment]
> 1. So it is worth while to continue a re-write? What would be conditions for a re-write to be accepted?

Yes, I think so. The quantum module needs a lot of love and you seem like someone who knows what you're doing. It's difficult to answer the question about accepting a change without seeing the code but I don't see any reason why a rewrite that fixes issues, improves things etc wouldn't be accepted.

> 2\. If yes - what about thinking to carry over qapply to these objects too - actually both packages do linear algebra?

I don't understand precisely what this proposal means. Can you give a concrete example of something that might be made possible by such a change?

> 3\. Pitfalls you can think of?

None from my side. Please be aware though that there are no maintainers who are significantly involved with the quantum module so it's possible that you now know more about its current state and issues than anyone else.
